### PR TITLE
THRIFT-3219: Provide a C++ tutorial on server-side IP logging and per-connection state

### PR DESCRIPTION
**Pull request by ben-craig**
This also adds the C++ tutorial builds to the cmake build by default.

The tutorial demonstrates per-connection state as a side effect.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -73,6 +73,7 @@ endif()
 
 if(BUILD_CPP)
     add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/lib/cpp)
+    add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/tutorial/cpp)
     if(BUILD_TESTING)
```
Comment (bufferoverflow) [inline CMakeLists.txt]: we probably need to add a BUILD_TUTORIALS option in the future

File: tutorial/cpp/CppLogIPServer.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    return new CalculatorHandler;
+  }
+  virtual void releaseHandler( ::shared::SharedServiceIf* handler) {
+    delete handler;
+  }
+};
+
+int main() {
+  TThreadedServer server(
+    boost::make_shared<CalculatorProcessorFactory>(boost::make_shared<CalculatorCloneFactory>()),
```
Comment (bufferoverflow) [inline tutorial/cpp/CppLogIPServer.cpp]: I would say we can add this to the regular Tutorial server and add some

Comment (ben-craig) [inline tutorial/cpp/CppLogIPServer.cpp]: Ok, combined the old and the new tutorials in the latest commit.

**Comment by bufferoverflow** [comment]
Thanks Ben!
Could you commit this?

**Pull request closed without merging**